File: components/common-go/experiments/configcat.go
// Copyright (c) 2022 Gitpod GmbH. All rights reserved.
// Licensed under the GNU Affero General Public License (AGPL).
// See License.AGPL.txt in the project root for license information.

package experiments

import (
	"context"
	"fmt"

	configcat "github.com/configcat/go-sdk/v7"
	"github.com/gitpod-io/gitpod/common-go/log"
	"github.com/sirupsen/logrus"
)

const (
	userIDAttribute         = "user_id"
	projectIDAttribute      = "project_id"
	teamIDAttribute         = "team_id"
	teamNameAttribute       = "team_name"
	vscodeClientIDAttribute = "vscode_client_id"
	gitpodHost              = "gitpod_host"
	component               = "component"
)

func newConfigCatClient(config configcat.Config) *configCatClient {
	return &configCatClient{
		client: configcat.NewCustomClient(config),
	}
}

var _ Client = (*configCatClient)(nil)

type configCatClient struct {
	client *configcat.Client
}

func (c *configCatClient) GetBoolValue(ctx context.Context, experimentName string, defaultValue bool, attributes Attributes) bool {
	value := c.client.GetBoolValue(experimentName, defaultValue, attributesToUser(attributes))
	log.AddFields(ctx, logField(experimentName, value))
	return value
}

func (c *configCatClient) GetIntValue(ctx context.Context, experimentName string, defaultValue int, attributes Attributes) int {
	value := c.client.GetIntValue(experimentName, defaultValue, attributesToUser(attributes))
	log.AddFields(ctx, logField(experimentName, value))
	return value
}

func (c *configCatClient) GetFloatValue(ctx context.Context, experimentName string, defaultValue float64, attributes Attributes) float64 {
	value := c.client.GetFloatValue(experimentName, defaultValue, attributesToUser(attributes))
	log.AddFields(ctx, logField(experimentName, value))
	return value
}

func (c *configCatClient) GetStringValue(ctx context.Context, experimentName string, defaultValue string, attributes Attributes) string {
	value := c.client.GetStringValue(experimentName, defaultValue, attributesToUser(attributes))
	log.AddFields(ctx, logField(experimentName, value))
	return value
}

func attributesToUser(attributes Attributes) *configcat.UserData {
	custom := make(map[string]string)

	if attributes.UserID != "" {
		custom[userIDAttribute] = attributes.UserID
	}

	if attributes.TeamID != "" {
		custom[teamIDAttribute] = attributes.TeamID
	}

	if attributes.TeamName != "" {
		custom[teamNameAttribute] = attributes.TeamName
	}

	if attributes.ProjectID != "" {
		custom[projectIDAttribute] = attributes.ProjectID
	}

	if attributes.VSCodeClientID != "" {
		custom[vscodeClientIDAttribute] = attributes.VSCodeClientID
	}

	if attributes.GitpodHost != "" {
		custom[gitpodHost] = attributes.GitpodHost
	}

	if attributes.Component != "" {
		custom[component] = attributes.Component
	}

	return &configcat.UserData{
		Identifier: attributes.UserID,
		Email:      attributes.UserEmail,
		Country:    "",
		Custom:     custom,
	}
}

type configCatLogger struct {
	*logrus.Entry
}

func (l *configCatLogger) GetLevel() configcat.LogLevel {
	return configcat.LogLevelError
}

func (l *configCatLogger) Debugf(format string, args ...interface{}) {}
func (l *configCatLogger) Infof(format string, args ...interface{})  {}
func (l *configCatLogger) Warnf(format string, args ...interface{})  {}

func logField(experimentName, value interface{}) logrus.Fields {
	return logrus.Fields{
		fmt.Sprintf("experiments.%s", experimentName): value,
	}
}

File: components/gitpod-protocol/src/experiments/configcat.ts
/**
 * Copyright (c) 2022 Gitpod GmbH. All rights reserved.
 * Licensed under the GNU Affero General Public License (AGPL).
 * See License.AGPL.txt in the project root for license information.
 */

import { Attributes, Client } from "./types";
import { User as ConfigCatUser } from "configcat-common/lib/RolloutEvaluator";
import { IConfigCatClient } from "configcat-common/lib/ConfigCatClient";

export const USER_ID_ATTRIBUTE = "user_id";
export const PROJECT_ID_ATTRIBUTE = "project_id";
export const TEAM_ID_ATTRIBUTE = "team_id";
export const TEAM_NAME_ATTRIBUTE = "team_name";
export const BILLING_TIER_ATTRIBUTE = "billing_tier";
export const GITPOD_HOST = "gitpod_host";

export class ConfigCatClient implements Client {
    constructor(private readonly client: IConfigCatClient, private readonly gitpodHost?: string) {}

    getValueAsync<T>(experimentName: string, defaultValue: T, attributes: Attributes): Promise<T> {
        return this.client.getValueAsync(
            experimentName,
            defaultValue,
            attributesToUser({
                gitpodHost: this.gitpodHost,
                ...attributes,
            }),
        );
    }

    dispose(): void {
        return this.client.dispose();
    }
}

export function attributesToUser(attributes: Attributes): ConfigCatUser {
    const userId = attributes.user?.id || "";
    const email = attributes.user?.email || "";

    const custom: { [key: string]: string } = {};
    if (userId) {
        custom[USER_ID_ATTRIBUTE] = userId;
    }
    if (attributes.projectId) {
        custom[PROJECT_ID_ATTRIBUTE] = attributes.projectId;
    }
    if (attributes.teamId) {
        custom[TEAM_ID_ATTRIBUTE] = attributes.teamId;
    }
    if (attributes.teamName) {
        custom[TEAM_NAME_ATTRIBUTE] = attributes.teamName;
    }
    if (attributes.billingTier) {
        custom[BILLING_TIER_ATTRIBUTE] = attributes.billingTier;
    }
    if (attributes.gitpodHost) {
        custom[GITPOD_HOST] = attributes.gitpodHost;
    }

    return new ConfigCatUser(userId, email, "", custom);
}

export interface ClassicPaygSunsetConfig {
    enabled: boolean;
    exemptedOrganizations: string[];
}
